kernel/src/net/socket/ip/raw/mod.rs

```diff
@@ -0,0 +1,427 @@
+            return_errno_with_message!(Errno::EINVAL, "the socket is not bound")
+        };
+        bound_raw.set_remote_addr(&endpoint.addr);
+
+        Ok(())
+    }
+
+    fn addr(&self) -> Result<SocketAddr> {
+        let inner = self.inner.read();
+        match inner.as_ref() {
+            Inner::Unbound(_) => Ok(UNSPECIFIED_LOCAL_ENDPOINT.into()),
+            Inner::Bound(bound_raw) => Ok(bound_raw.local_endpoint().into()),
+        }
+    }
+
+    fn peer_addr(&self) -> Result<SocketAddr> {
+        self.remote_addr()
+            .map(|addr| IpEndpoint::new(addr, 0).into())
+            .ok_or_else(|| Error::with_message(Errno::ENOTCONN, "the socket is not connected"))
+    }
+
+    fn sendmsg(
+        &self,
+        reader: &mut dyn MultiRead,
+        message_header: MessageHeader,
+        flags: SendRecvFlags,
+    ) -> Result<usize> {
+        // TODO: Deal with flags
+        debug_assert!(flags.is_all_supported());
+
+        let MessageHeader {
+            addr,
+            control_message,
+        } = message_header;
+
+        let remote_endpoint = match addr {
+            Some(remote_addr) => {
+                let endpoint = remote_addr.try_into()?;
+                self.try_bind_ephemeral(&endpoint)?;
+                endpoint
+            }
+            None => {
+                let ip_addr: IpEndpoint = match self.addr() {
+                    Ok(ip_addr) => ip_addr.try_into()?,
+                    Err(_) => {
+                        return Err(Error::with_message(
+                            Errno::EDESTADDRREQ,
+                            "the destination address is not specified",
+                        ))
+                    }
+                };
+                IpEndpoint::new(ip_addr.addr, 0)
+            }
+        };
+
+        if control_message.is_some() {
+            // TODO: Support sending control message
+            warn!("sending control message is not supported");
+        }
+
+        let mut op = IpHdrIncl::new();
+        let _ = self.get_option(&mut op);
+
+        // TODO: Block if the send buffer is full
+        self.try_send(reader, &remote_endpoint.addr, flags, op)
+    }
+
+    fn recvmsg(
+        &self,
+        writer: &mut dyn MultiWrite,
+        flags: SendRecvFlags,
+    ) -> Result<(usize, MessageHeader)> {
+        // TODO: Deal with flags
+        debug_assert!(flags.is_all_supported());
+
+        // This is a temporary solution. Bind to a specific iface: iface[0]
+        // TODO: Monitor multiple ifaces
+        let bind_iface_addr = IpEndpoint::new(IpAddress::v4(10, 0, 2, 15), 0);
+        self.try_bind_ephemeral(&bind_iface_addr)?;
+
+        let (received_bytes, addr) = self.recv(writer, flags)?;
+        // TODO: Receive control message
+        let IpAddress::Ipv4(ip_addr) = addr;
+        let message_header = MessageHeader::new(Some(SocketAddr::IPv4(ip_addr, 0)), None);
+        Ok((received_bytes, message_header))
```

Comment:
Add blank lines to split unrelated code blocks. I think they're required:
 - before `TODO: Receive control message`
 - after `TODO: Receive control message`
 - after `let message_header = ...`